shared/src/event_handling.rs

```diff
@@ -0,0 +1,170 @@
+use anyhow::{Context, Error, Result};
+use ethcontract::contract::{AllEventsBuilder, ParseLog};
+use ethcontract::errors::ExecutionError;
+use ethcontract::{dyns::DynTransport, BlockNumber as Web3BlockNumber, Event as EthcontractEvent};
+use futures::{Stream, StreamExt, TryStreamExt};
+use std::ops::RangeInclusive;
+use web3::Web3;
+
+// We expect that there is never a reorg that changes more than the last n blocks.
+const MAX_REORG_BLOCK_COUNT: u64 = 25;
+// When we insert new trade events into the database we will insert at most this many in one
+// transaction.
+const INSERT_EVENT_BATCH_SIZE: usize = 10_000;
+
+pub struct EventHandler<C: EventRetrieving, S: EventStoring<C::Event>> {
+    contract: C,
+    store: S,
+    last_handled_block: Option<u64>,
+}
+
+#[async_trait::async_trait]
+pub trait EventStoring<T> {
```

Comment:
Maybe some trait documentation might be helpful here (its obvious to me, but I'm biased because we had discussed it already :stuck_out_tongue:). An example of something that might not be obvious at first glance is the optional `range` parameter in `save_events`.